File: historical/milinda/src/org/openexi/proc/io/Characters.java
package org.openexi.proc.io;

import org.openexi.proc.common.CharacterSequence;

final class Characters implements CharacterSequence {

  private final char[] m_characters;
  private final int m_startIndex;
  private final int m_length;
  private final int m_mojiCount;
  private final int m_hashCode;

  public static final CharacterSequence TRUE;
  public static final CharacterSequence FALSE;
  static {
    TRUE = new Characters("true".toCharArray(), 0, "true".length());
    FALSE = new Characters("false".toCharArray(), 0, "false".length());
  }
  
  Characters(char[] characters, int startIndex, int length) {
    m_characters = characters;
    m_startIndex = startIndex;
    m_length = length;
    int hashCode = 0;
    final int limit = m_startIndex + m_length;
    int mojiCount = 0;
    for (int i = m_startIndex; i < limit; ++mojiCount) {
      char currentChar = m_characters[i++];
      hashCode = hashCode * 31 + currentChar;
      if ((currentChar & 0xFC00) == 0xD800) { // high surrogate
        if (i < limit) {
          currentChar = m_characters[i];
          if ((currentChar & 0xFC00) == 0xDC00) { // low surrogate
            ++i;
          }
        }
      }
    }
    m_mojiCount = mojiCount;
    m_hashCode = hashCode;
  }
  
  public char[] getCharacters() {
    return m_characters;
  }
  
  public int getStartIndex() {
    return m_startIndex;
  }
  
  public int length() {
    return m_length;
  }
  
  public int getUCSCount() {
    return m_mojiCount;
  }

  public int indexOf(char c) {
    final int limit = m_startIndex + m_length;
    for (int i = m_startIndex; i < limit; i++) {
      if (m_characters[i] == c) 
        return i;
    }
    return -1;
  }

  public String substring(int beginIndex, int endIndex) {
    return new String(m_characters, beginIndex, endIndex - beginIndex);
  }

  @Override
  public int hashCode() {
    return m_hashCode;
  }

  @Override
  public boolean equals(Object object) {
    if (object instanceof Characters) {
      final Characters characterSequence = (Characters)object;
      if (m_length != characterSequence.m_length) 
        return false;
      else {
        final int startIndex = characterSequence.m_startIndex;
        final char[] characters = characterSequence.m_characters;
        for (int i = 0; i < m_length; i++) {
          if (m_characters[m_startIndex + i] != characters[startIndex + i])
            return false;
        }
        return true;
      }
    }
    else
      return false;
  }
  
  public String makeString() {
    return new String(m_characters, m_startIndex, m_length);
  }

}

File: nagasenacs/src/Nagasena/Schema/Characters.cs
using System;

namespace Nagasena.Schema {

  public sealed class Characters {

    public bool isVolatile;

    public char[] characters;
    public int startIndex;
    public readonly int length;
    public readonly int ucsCount;
    private readonly int m_hashCode;

    public static readonly Characters CHARACTERS_EMPTY;
    static Characters() {
      CHARACTERS_EMPTY = new Characters("".ToCharArray(), 0, 0, false);
    }

    public Characters(char[] characters, int startIndex, int length, bool isVolatile) {
      this.isVolatile = isVolatile;
      this.characters = characters;
      this.startIndex = startIndex;
      this.length = length;
      int hashCode = 0;
      int limit = this.startIndex + this.length;
      int mojiCount = 0;
      for (int i = this.startIndex; i < limit; ++mojiCount) {
        char currentChar = this.characters[i++];
        hashCode = hashCode * 31 + currentChar;
        if ((currentChar & 0xFC00) == 0xD800) { // high surrogate
          if (i < limit) {
            currentChar = this.characters[i];
            if ((currentChar & 0xFC00) == 0xDC00) { // low surrogate
              ++i;
            }
          }
        }
      }
      this.ucsCount = mojiCount;
      m_hashCode = hashCode;
    }

    public void turnPermanent() {
      if (isVolatile) {
        char[] _characters = new char[length];
        Array.Copy(characters, startIndex, _characters, 0, length);
        characters = _characters;
        startIndex = 0;
        isVolatile = false;
      }
      return;
    }

    public int IndexOf(char c) {
      int limit = startIndex + length;
      for (int i = startIndex; i < limit; i++) {
        if (characters[i] == c) {
          return i;
        }
      }
      return -1;
    }

    public string Substring(int beginIndex, int endIndex) {
      return new string(characters, beginIndex, endIndex - beginIndex);
    }

    public override int GetHashCode() {
      return m_hashCode;
    }

    public override bool Equals(object @object) {
      if (@object is Characters) {
        Characters characterSequence = (Characters)@object;
        if (length != characterSequence.length) {
          return false;
        }
        else {
          int startIndex = characterSequence.startIndex;
          char[] characters = characterSequence.characters;
          for (int i = 0; i < length; i++) {
            if (this.characters[this.startIndex + i] != characters[startIndex + i]) {
              return false;
            }
          }
          return true;
        }
      }
      else {
        return false;
      }
    }

    public string makeString() {
      return new string(characters, startIndex, length);
    }

  }

}